udaykalra/walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,133 @@
 // limitations under the License.
 
 package com.google.sps;
-
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 
+/**
+ * Returns a collection of TimeRange objects to indicate available meeting times
+ * for indicated people and duration given through a request object.
+ */
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    /* Create an output collection with a full day to cut(shorten or split) 
+     * into the available ranges.
+     */
+    TimeRange fullDay = TimeRange.WHOLE_DAY;
+    Collection<TimeRange> availableRanges = new HashSet<>();
+    availableRanges.add(fullDay);
+
+    // Obtain data from the request.
+    Collection<String> attendees = request.getAttendees();
+    long requestDuration = request.getDuration();
+
+    Collection<TimeRange> busyTimes = new HashSet<>();
+
+    // Detect relevant busy time ranges based on attendees.
+    for (Event event : events) {
+      boolean attendeePresent = false;
+      if (!(Collections.disjoint(attendees, event.getAttendees()))) {
+        busyTimes.add(event.getWhen());
+      }
+    }
+    
+
+    /* Iteratively cut availableRange's times based on 
+     * overlap with busy events and store results in availableRanges.
+     */
+    for (TimeRange busy : busyTimes) {
```

Comment:
Suggest naming `busyTime`, to make it easier to read below (in general readability is preferred over saving horizontal space).

Reply:
Done